### PR TITLE
[SPARK-17702][SQL] Code generation including too many mutable states exceeds JVM size limit.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

Code generation including too many mutable states exceeds JVM size limit to extract values from `references` into fields in the constructor.
We should split the generated extractions in the constructor into smaller functions.
## How was this patch tested?

I added some tests to check if the generated codes for the expressions exceed or not.

### CODE REVIEW

**Comment by rxin** [comment]
Can you add more inline comments explaining what's going on?

**Comment by rxin** [comment]
LGTM. cc @davies to take a look too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -589,6 +592,11 @@ class CodegenContext {
       // Cannot split these expressions because they are not created from a row object.
       return expressions.mkString("\n")
     }
+    splitExpressions(expressions, "apply", ("InternalRow", row) :: Nil)
+  }
+
+  private def splitExpressions(
+      expressions: Seq[String], funcName: String, arguments: Seq[(String, String)]): String = {
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: How about exposing this API as non-public? This API looks more flexible and reusable.Is there any reason to declare as private?

Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I didn't have a special reason.
Should we make it public? @rxin, @davies

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: As far as I know, there are two possible use cases of this API
https://issues.apache.org/jira/browse/SPARK-16223
https://github.com/apache/spark/pull/15219/files#diff-8bcc5aea39c73d4bf38aef6f6951d42cL587 in https://github.com/apache/spark/pull/15219

**Comment by rxin** [comment]
Merging in master. Thanks.

**Comment by viirya** [comment]
This affects many places in code generation. I think this should be a common issue for the users of previous versions. Should we make a backport to branch-2.0 and branch-1.6? cc @rxin @ueshin @kiszk

**Comment by viirya** [comment]
One important reason to prepare a backport for this is, as we deprecate the settings to turn off unsafe and codegen since 1.6, we can't turn them off to avoid JVM size limitation issue if users encounter that.

**Comment by ueshin** [comment]
It seems we can cherry-pick into `branch-2.0` without conflicts, but can't into `branch-1.6`.
Please let me know if we need to backport to `branch-1.6`.

**Comment by rezasafi** [comment]
Any update on the decision of whether this fix will be backported to 1.6? thanks.

**Comment by xs2rajni** [comment]
When will this fix be backported to spark 2.0?

**Pull request closed without merging**